### PR TITLE
Initial vision API addition.

### PR DESCRIPTION
**Pull request by daspecster**
Fairly rough and a lot of conversation to be had.

- [ ] Face
- [ ] Landmark
- [ ] Logo
- [ ] Label
- [ ] Text
- [ ] Safe Search

Examples to play with
```python
>>> from gcloud.vision import Client
>>> vc = Client('ferrous-arena')
>>> f = open('/tmp/car.jpg', 'r')
>>> vc.annotate(f.read(), 'LABEL_DETECTION', 3)
>>> vc.annotate('gs://ferrous-arena-my-test-bucket/car.jpg', 'LABEL_DETECTION', 3)
```

### CODE REVIEW

File: gcloud/vision/annotation.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+class EntityAnnotation(object):
+    _mid = None
+    _locale = None
+    _description = None
+    _score = None
+    _confidence = None
+    _topicality = None
+    _bounding_poly = None
+    _locations = []
+    _properties = []
```
Comment (dhermes) [inline gcloud/vision/annotation.py]: Mutable class attributes?

File: gcloud/vision/client.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import json
+from json import JSONEncoder
+
+from gcloud.client import JSONClient
+from gcloud.vision.connection import Connection
+from gcloud.vision.feature import Feature
+from gcloud.vision.image import Image
+
+
+class VisionJSONEncoder(JSONEncoder):
```
Comment (dhermes) [inline gcloud/vision/client.py]: This seems like it'll be overkill (haven't seen how you're using it though).

File: gcloud/vision/client.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        else:
+            return o.__dict__
+
+
+class VisionRequest(object):
+    def __init__(self, image, feature):
+        self._features = []
+        self._image = image
+
+        if type(feature) == list:
```
Comment (dhermes) [inline gcloud/vision/client.py]: Use `isinstance`

**Comment by jonparrott** [comment]
```
vc.annotate(f.read(), 'LABEL_DETECTION', 3)
```
I'd prefer constants over strings, even if they semantically mean the same thing:

```
vc.annotate(f.read(), vision.LABEL_DETECTION, 3)
```

File: gcloud/vision/client.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            self._features.extend(feature)
+        elif type(feature) == Feature:
+            self._features.append(feature)
+
+    def as_dict(self):
+        return {
+            "image": self.image,
+            "features": self.features
```
Comment (dhermes) [inline gcloud/vision/client.py]: Single quotes

File: gcloud/vision/client.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                        owned by this client. If not passed (and if no ``http``
+                        object is passed), falls back to the default inferred
+                        from the environment.
+
+    :type http: :class:`httplib2.Http` or class that defines ``request()``.
+    :param http: An optional HTTP object to make requests. If not passed, an
+                 ``http`` object is created that is bound to the
+                 ``credentials`` for the current object.
+
+    Usage:
```
Comment (dhermes) [inline gcloud/vision/client.py]: You either need a Sphinx code-block or `::` to initiate one

File: gcloud/vision/client.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    >>> from gcloud import vision
+    >>> vision_client = vision.Client()
+    >>> f = open('/tmp/car.jpg', 'r')
+    >>> vision_client.annotate(f.read(), "LABEL_DETECTION", 3)
```
Comment (dhermes) [inline gcloud/vision/client.py]: Example should use a `file` as a context manager.

File: gcloud/vision/client.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+    Multiple images example:
+    >>> images = (("./image.jpg", ["LABEL_DETECTION", "LANDMARK_DETECTION"]),
+    >>>           ("./image2.jpg", ["FACE_DETECTION", "TEXT_DETECTION"]),)
+    >>> annotated_images = []
+    >>> for image, feature_types in images:
+    >>>     annotated_images.append(vision_client.annotate(image,
+    >>>                                                    feature_types))
```
Comment (dhermes) [inline gcloud/vision/client.py]: `...` instead of `>>>`

File: gcloud/vision/client.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                 ``http`` object is created that is bound to the
+                 ``credentials`` for the current object.
+
+    Usage:
+    >>> from gcloud import vision
+    >>> vision_client = vision.Client()
+    >>> f = open('/tmp/car.jpg', 'r')
+    >>> vision_client.annotate(f.read(), "LABEL_DETECTION", 3)
+
+    Multiple images example:
```
Comment (dhermes) [inline gcloud/vision/client.py]: Ditto here on the code block

File: gcloud/vision/feature.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    See:
+    https://cloud.google.com/vision/reference/rest/v1/images/annotate#Type
+    """
+    FACE_DETECTION = "FACE_DETECTION"
+    LANDMARK_DETECTION = "LANDMARK_DETECTION"
+    LOGO_DETECTION = "LOGO_DETECTION"
+    LABEL_DETECTION = "LABEL_DETECTION"
+    TEXT_DETECTION = "TEXT_DETECTION"
+    SAFE_SEARCH_DETECTION = "SAFE_SEARCH_DETECTION"
```
Comment (dhermes) [inline gcloud/vision/feature.py]: Single quotes, Single quotes, Single quotes

File: gcloud/vision/feature.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+
+class Feature(object):
+    """Feature object specifying the annotation and number of results for the
+    specified feature type.
```
Comment (dhermes) [inline gcloud/vision/feature.py]: One-line summary should be one line.

File: gcloud/vision/client.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            if type(feature_type) == list:
+                for feature in feature_type:
+                    features.append(Feature(feature, max_results))
+            else:
+                features.append(Feature(feature_type, max_results))
+
+            vision_requests.append(VisionRequest(img, features))
+
+        data = {"requests": vision_requests}
+        data = json.dumps(data, cls=VisionJSONEncoder)
```
Comment (dhermes) [inline gcloud/vision/client.py]: Could've you just do:

```python
data = {'request': [req.as_dict() for req in vision_requests]}
```

or even better, don't build up `vision_requests` when you only care about building up a list of

```python
VisionRequest(img, features).as_dict()
```

File: gcloud/vision/image.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+    See:
+    https://cloud.google.com/vision/reference/rest/v1/images/annotate#Image
+    """
+    def __init__(self, image):
+        self._source = None
+        self._content = None
+
+        # TODO: string index out of range for len(string) < 4.
+        # TODO: Support file path.
```
Comment (dhermes) [inline gcloud/vision/image.py]: remove TODO -> file an issue

File: gcloud/vision/image.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        if image[:4] == 'http':
+            import httplib2
```
Comment (dhermes) [inline gcloud/vision/image.py]: Imports at top-level

File: gcloud/vision/image.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    See:
+    https://cloud.google.com/vision/reference/rest/v1/images/annotate#Image
+    """
+    def __init__(self, image):
+        self._source = None
+        self._content = None
+
+        # TODO: string index out of range for len(string) < 4.
+        # TODO: Support file path.
+        if image[:4] == 'http':
```
Comment (dhermes) [inline gcloud/vision/image.py]: What's wrong with `startswith`?

File: gcloud/vision/test_client.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
...0gX9s+5L//9rd1P/dfPL98NdBoRx22NQwsLLDySUvtoAbi4opERppKE9UnCCFD7gx+tPfHfXzLywq0zwTNrI3JOpW5JreAj4zOulxGTg907pNpk7RTECoIpUXRLQ5XhUOOgO6Rk6lh6Ry9mjdw0VbzzSuJVEzCxM4GDbtPRbumh+46mX//o7a3P/9fPtb8HtMOQVhhysNVijrMYdnysAlxcI+snmTk4zzQwCgSvfubI1qsvja560vroqrERLmg07IiEhiQAHxiVvOxUglyyc96X2oxtpaWWD8NH62iJICd5K0nx/T1XM79W1Xk0BZN6CfNU3/Kyay222PXdg/FNO74T3/T+r7fuy5X80vpJjh2c5ygPs/vzWAQYQNbBSFJjcq7NDNAAuGxjtPaXrojOf855lcvOGTOXjo3I1rBqRiqRwYeCDwQ1glpRJFfhGXdLxEg/czUkwUVrBHzW+Esd4D3iVMQrJlVMonRjT9LxrcWW3ndo0X/ny7u73/7QzfE93z4Qz+afuDxV42jYZv5wBqw+aiaUR+clQHVkhAlJmG4m1PNdtPGsbXbspRdVN16xubJ984Q8abJhtoSRmQgDGYsqYjTvYKsmy26AqJqsF2C5hg9FxIuAivHaO59KnCfuqk9SXUxif3x+2e/dd1y/e/O+7s5P3tU58LU9bjEPK/p6SFNDjrZaLOQ+rT4aJ/LRfoXj4zSqMDG3wGgCVbI7AnjpReH0c7dVz902Y9dvnpat0w27YSQyU9VQJwNrxquRRGFQ6g1CPxKVpEon1jh1fqGTyHwr9nPHlt0D+47pfXuOuoNf2tO5/5N3JQXhLQY6IXSmxlnqwPGFBZYfKX/2+xngntW9BaL5MUZsTL3ZYTSGkRxsITs0fOQFF1RnzpuWyXMadmxqRBqjNVMdCaRaCQkDm/FXU6e+m5C0Uu0stX1nrqXLh5bd4u5jOv+FXZ2jrZhWyTvuRNCqV1lyEc3JRVp7M7D9Y2HSHksADz93AISjUEtr1AJHbSmmSv9U7JVofA9WilZ4yMloRCe1tIM27SVo088T6WNxor5frmAThNUpQmIq7YhK4AhcRBB4rCrGa5id8yKJF8GnBmdj0tSS1mK6RHQ7cyT7+4A+5q/vJ4BXGttJeLAnRKL1sSidj1+PX49fj1+PX49fj1+P3ut/A6a1MamkZwQCAAAAAElFTkSuQmCC"
```
Comment (dhermes) [inline gcloud/vision/test_client.py]: 1. You have `IMAGE_CONTENT` twice
1. Do you really need a real image for a unit test?
1. If you actually do, put it in a file.

**Comment by dhermes** [comment]
Rather than writing and dumping all this code, please draft a `vision-usage.rst` doc and we can discuss features before there is code.

Once we agree on something, try to land them feature by feature so review doesn't become unmanageable. (e.g. you couldn't just sent a tiny PR with the `Connection`, then a tiny PR with the `Client` and then build up the client for each feature).

(PS I know I "broke the rules" in #1772 however the entire API is contained in three methods so I did it "on purpose" / was aware it was a bit of a brain dump.)

**Comment by tseaver** [comment]
> (PS I know I "broke the rules" in #1772 however the entire API is contained in three methods so I did it "on purpose" / was aware it was a bit of a brain dump.)

#1772 has have 70+ comments on it, which is a sign that we should've argued out the API surface first. :)

File: gcloud/vision/client.py (hunk 4 above)
Comment (daspecster) [inline gcloud/vision/client.py]: Thanks, flipping between JSON makes my shift finger twitchy I guess. I think I got them all in the next commit.

File: gcloud/vision/client.py (hunk 5 above)
Comment (daspecster) [inline gcloud/vision/client.py]: good call, I haven't tried to generate docs for this yet.

File: gcloud/vision/client.py (hunk 6 above)
Comment (daspecster) [inline gcloud/vision/client.py]: 👍

**Comment by dhermes** [comment]
@tseaver I totally agree and wish I had done that.

File: gcloud/vision/feature.py (hunk 9 above)
Comment (daspecster) [inline gcloud/vision/feature.py]: /me ties pinky fingers back with rubberband.

File: gcloud/vision/client.py (hunk 11 above)
Comment (daspecster) [inline gcloud/vision/client.py]: 👍

File: gcloud/vision/image.py (hunk 14 above)
Comment (daspecster) [inline gcloud/vision/image.py]: not a thing! fixed.

File: gcloud/vision/test_client.py (hunk 15 above)
Comment (daspecster) [inline gcloud/vision/test_client.py]: I don't, I was testing a bug and I unfortunately forgot it.

**Comment by daspecster** [comment]
Ok, I'll make a PR with just a vision-usage.rst and then we can talk about it there.

**Comment by tseaver** [comment]
In favor of #1809 ans subsequent.

**Pull request closed without merging**